influxdb3_write/src/last_cache.rs

```diff
@@ -0,0 +1,2225 @@
+    }
+}
+
+/// Represents the hierarchical last cache structure
+#[derive(Debug)]
+enum LastCacheState {
+    /// An initialized state that is used for easy construction of the cache
+    Init,
+    /// Represents a branch node in the hierarchy of key columns for the cache
+    Key(LastCacheKey),
+    /// Represents a terminal node in the hierarchy, i.e., the cache of field values
+    Store(LastCacheStore),
+}
+
+impl LastCacheState {
+    fn is_init(&self) -> bool {
+        matches!(self, Self::Init)
+    }
+
+    fn as_key(&self) -> Option<&LastCacheKey> {
+        match self {
+            LastCacheState::Key(key) => Some(key),
+            LastCacheState::Store(_) | LastCacheState::Init => None,
+        }
+    }
+
+    fn as_store(&self) -> Option<&LastCacheStore> {
+        match self {
+            LastCacheState::Key(_) | LastCacheState::Init => None,
+            LastCacheState::Store(store) => Some(store),
+        }
+    }
+
+    fn as_key_mut(&mut self) -> Option<&mut LastCacheKey> {
+        match self {
+            LastCacheState::Key(key) => Some(key),
+            LastCacheState::Store(_) | LastCacheState::Init => None,
+        }
+    }
+
+    fn as_store_mut(&mut self) -> Option<&mut LastCacheStore> {
+        match self {
+            LastCacheState::Key(_) | LastCacheState::Init => None,
+            LastCacheState::Store(store) => Some(store),
+        }
+    }
+
+    /// Remove expired values from this [`LastCacheState`]
+    fn remove_expired(&mut self) {
+        match self {
+            LastCacheState::Key(k) => k.remove_expired(),
+            LastCacheState::Store(s) => s.remove_expired(),
+            LastCacheState::Init => (),
+        }
+    }
+}
+
+/// Holds a node within a [`LastCache`] for a given key column
+#[derive(Debug)]
+struct LastCacheKey {
+    /// The name of the key column
+    column_name: String,
+    /// A map of key column value to nested [`LastCacheState`]
+    ///
+    /// All values should point at either another key or a [`LastCacheStore`]
+    value_map: HashMap<KeyValue, LastCacheState>,
+}
+
+impl LastCacheKey {
+    /// Evaluate the provided [`Predicate`] by using its value to lookup in this [`LastCacheKey`]'s
+    /// value map.
+    ///
+    /// # Panics
+    ///
+    /// This assumes that a predicate for this [`LastCacheKey`]'s column was provided, and will panic
+    /// otherwise.
+    fn evaluate_predicate(&self, predicate: &Predicate) -> Option<&LastCacheState> {
+        if predicate.key != self.column_name {
+            panic!(
+                "attempted to evaluate unexpected predicate with key {} for column named {}",
+                predicate.key, self.column_name
+            );
+        }
+        self.value_map.get(&predicate.value)
+    }
+
+    /// Remove expired values from any cache nested within this [`LastCacheKey`]
+    fn remove_expired(&mut self) {
+        self.value_map
+            .iter_mut()
+            .for_each(|(_, m)| m.remove_expired());
```

Comment:
If all rows have been expired, this key should be removed from the map. Otherwise, key values that stop sending data (think ephemeral things like container id, etc) will blow up the size of the cache over time with a bunch of entry map entries. (I'm assuming I'm reading this correctly and only the values are getting removed).

This should be true walking up the tree. Each key/value should be removed from the map if all children are empty.

Reply:
Yes, good call, this only removes the values and is not walking up and cleaning up the maps. I'll address that with the other immediate issues in a follow-on PR.